### PR TITLE
Engine moved to a Class Library.

### PR DESCRIPTION
**Pull request by naicigam**
Now the Engine can be used from an custom VS project without having to open the entire solution and adding a new project.

### CODE REVIEW

File: Launcher/Program.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+namespace Launcher
+{
+    class Program
+    {
+        /// <summary>
+        /// Primary Analysis Thread:
+        /// </summary>
+        public static void Main(string[] args)
+        {
+            QuantConnect.Lean.Engine.Engine.Launch();
```
Comment (jaredbroad) [inline Launcher/Program.cs]: The Engine should take config and run as an instance - I'd recommend moving the entire of the Main() method into the Launcher Main(). The call to engine should look something like this:

```
//Get the configuration
var config = ...;
//Launch the engine
var engine = new Engine(config);
engine.Run();
```

Comment (naicigam) [inline Launcher/Program.cs]: Great! I will do it that way.

Comment (naicigam) [inline Launcher/Program.cs]: @jaredbroad Just started to make this change and I noticed that Config is static, so what would be the point of passing it to the Engine? In that case, it seems OK for the Engine to retrieve the values in the Constructor. Something like:

```
namespace Launcher
{
    class Program
    {
        public static void Main(string[] args)
        {
                var engine = new Engine();
                engine.Run();
        }
}
```

The other option is to get the configuration values in the Laucher.Main and passing LeanEngineSystemHandlers, LeanEngineAlgorithmHandlers and LiveMode to the Engine.
The problem I see with that is that the initialization code will have to be replicated in the application that uses the Engine DLL.

The third option involves many code changes: making the Config class non-static. But we will have to pass the Config object around to all the other classes that use it (ex. LiveTradingDataFeed).

What do you think?

Comment (jaredbroad) [inline Launcher/Program.cs]: Engine should take its configuration items in its constructor. This is what I meant by the "var config ...." line. Its actually the LeanEngineSystemHandlers etc. Basically everything in the Main() method can be moved into the Launcher (i.e. not duplicated, but moved from the Engine.Main()).

**Comment by bizcad** [comment]
@naicigam I like the idea of running Lean as a dll very much.  

In regards to pushing the Config into the dll, you might want to take a look at <a href="https://github.com/QuantConnect/Lean/issues/190"> issue #190 Genetic Optimization</a>.  @chrisdk2015 has a reference to running Lean from another project.  There he is passing parameters into the Config and setting LeanEnginexxxxxxHandlers.  He then runs Lean in a console with something like...

<code>
public decimal Run(int val)
        {
            Config.Set("EMA_VAR1", val.ToString(CultureInfo.InvariantCulture));
            LaunchLean();
            // Handle results
        }
</code>
and in the LaunchLean method...
<code>
        private void LaunchLean()
        {
            Config.Set("environment", "backtesting");
            string algorithm = "EMATest";  // tell Lean what algorithm to run.

            Config.Set("algorithm-type-name", algorithm);
            //string datapath = Config.Get("data-folder");
            _jobQueue = new JobQueue();
            _notify = new Messaging();
            _api = new Api();
            _resultshandler = new OptimizationResultHandler();
            //_resultshandler = new ConsoleResultHandler();
            _dataFeed = new FileSystemDataFeed();
            _setup = new ConsoleSetupHandler();
            _realTime = new BacktestingRealTimeHandler();
            _historyProvider = new SubscriptionDataReaderHistoryProvider();
            _transactions = new BacktestingTransactionHandler();
            Log.LogHandler = (ILogHandler)new FileLogHandler();
            Log.DebuggingEnabled = false;
            Log.DebuggingLevel = 1;

            var systemHandlers = new LeanEngineSystemHandlers(_jobQueue, _api, _notify);
            systemHandlers.Initialize();

            var algorithmHandlers = new LeanEngineAlgorithmHandlers(_resultshandler, _setup, _dataFeed, _transactions, _realTime, _historyProvider);
            string algorithmPath;
            
            AlgorithmNodePacket job = systemHandlers.JobQueue.NextJob(out algorithmPath);
            try
            {
                var _engine = new Engine(systemHandlers, algorithmHandlers, Config.GetBool("live-mode"));
                _engine.Run(job, algorithmPath);
            }
            finally
            {
                //Delete the message from the job queue:
                //systemHandlers.JobQueue.AcknowledgeJob(job);
                Log.Trace("Engine.Main(): Packet removed from queue: " + job.AlgorithmId);

                // clean up resources
                systemHandlers.Dispose();
                algorithmHandlers.Dispose();
                Log.LogHandler.Dispose();
            }

        }
</code>

You can find my version of the Optimization project in a <a href="https://github.com/bizcad/LeanOpto/blob/master/Optimization/Program.cs"> repo  on 
github.  </a>

On my local machine I have been using a version of his code to run various optimizations and have replaced ConsoleResultHandler with a custom OptimizationResultHandler.  I also have been running multiple algorithms and for various date ranges from a list in Program.cs like this...
<code>
        private static double RunAlgorithm(List<string> algos, Dictionary<string, DateRange> daysDictionary)
        {

            var sum_sharpe = 0.0;
            foreach (string s in algos)
            {
                foreach (string key in daysDictionary.Keys)
                {
                    var val = s;
                    var startDate = daysDictionary[key].startDate;
                    var endDate = daysDictionary[key].endDate;
                    AppDomain ad = null;
                    RunClass rc = CreateRunClassInAppDomain(ref ad);
                    Console.WriteLine("Running algorithm {0} for: {1} to {2}", val, startDate, endDate);

                    try
                    {
                        sum_sharpe += (double)rc.Run(val, startDate, endDate);
                    }
                    catch (Exception e)
                    {
                        Log.Error(e.Message + e.StackTrace);
                    }
                    AppDomain.Unload(ad);
          }
</code>
and 
<code>
        public decimal Run(string val, string startDate, string endDate)
        {
            // Set dates in Config.
            Config.Set("start-date", startDate);
            Config.Set("end-date", endDate);

            LaunchLean(val);   // pass in the algo name
...
       }
</code>
Then in the Initialize() of the algorithm...
<code>
            //Initialize dates
            sd = Config.Get("start-date");
            ed = Config.Get("end-date");

            _startDate = new DateTime(Convert.ToInt32(sd.Substring(0, 4)), Convert.ToInt32(sd.Substring(4, 2)), Convert.ToInt32(sd.Substring(6, 2)));
            _endDate = new DateTime(Convert.ToInt32(ed.Substring(0, 4)), Convert.ToInt32(ed.Substring(4, 2)), Convert.ToInt32(ed.Substring(6, 2)));


            SetStartDate(_startDate);
            SetEndDate(_endDate);
            SetCash(_portfolioAmount);
</code>

This code might give you some ideas on handling passing variables into the dll by making entries into Config.  

I hope this helps.  I am very much looking forward to your getting the Engine running as a dll.

Nick Stein aka bizcad

**Comment by mchandschuh** [comment]
Extracting the `Config` static class can be handled in another issue. I've added a QuantConnect.Lean.Launcher csproj that will now act as the start up project for the engine in console mode. The QuantConnect.Lean.Engine has been converted into a dll project.

Here's the [commit](https://github.com/QuantConnect/Lean/commit/681cd32e328a621f7c73e0bababa40330185207f).

**Pull request closed without merging**